File: test-fixtures/shape/systems/sample.cs
// Hand-written control-flow sample for the C# body-shape descriptor coverage
// test. Exercises branch, loop, switch, try/catch/throw, return, break/continue,
// a call, and a lambda so the canonical CfBucket histogram is non-empty.

using System;

namespace Shapes
{
    public class Classifier
    {
        public int Classify(int n, string label = "default")
        {
            int total = 0;
            Func<int, int> twice = x => x * 2;

            if (n > 0)
            {
                total = twice(n);
            }
            else
            {
                total = 0;
            }

            for (int i = 0; i < n; i++)
            {
                if (i == 3)
                {
                    continue;
                }
                total += i;
            }

            while (total < 100)
            {
                total += 1;
                if (total == 50)
                {
                    break;
                }
            }

            switch (n)
            {
                case 0:
                    Console.WriteLine(total);
                    break;
                default:
                    Console.WriteLine(0);
                    break;
            }

            try
            {
                if (total < 0)
                {
                    throw new InvalidOperationException("negative");
                }
            }
            catch (Exception)
            {
                total = -1;
            }

            return total;
        }
    }
}

File: test-fixtures/shape/systems/sample.scala
// Hand-written control-flow sample for the Scala body-shape descriptor coverage
// test. Exercises branch, loop, match, try/catch, throw, return, and a call so
// the canonical CfBucket histogram is non-empty.

object Classifier {
  def compute(value: Int): Int = value + 1

  def classify(n: Int, label: String): Int = {
    var total = 0
    if (n > 0) {
      total = compute(n)
    } else {
      total = 0
    }

    while (total < 100) {
      total = total + 1
    }

    val tag = n match {
      case 0 => "zero"
      case _ => "other"
    }

    try {
      if (tag == "zero") {
        throw new RuntimeException(label)
      }
    } catch {
      case _: RuntimeException => total = -1
    }

    return total
  }
}
